src/wallet/wallet2.cpp

```diff
@@ -5025,6 +5025,230 @@ bool wallet2::get_tx_key(const crypto::hash &txid, crypto::secret_key &tx_key) c
   tx_key = i->second;
   return true;
 }
+//----------------------------------------------------------------------------------------------------
+std::vector<std::vector<crypto::signature>> wallet2::get_spend_proof(const cryptonote::transaction_prefix& tx)
+{
+  THROW_WALLET_EXCEPTION_IF(m_watch_only, error::wallet_internal_error,
+    "get_spend_proof requires spend secret key and is not available for a watch-only wallet");
+
+  std::vector<std::vector<crypto::signature>> signatures;
+
+  const crypto::hash tx_prefix_hash = cryptonote::get_transaction_prefix_hash(tx);
+
+  for(size_t i = 0; i < tx.vin.size(); ++i)
+  {
+    if(tx.vin[i].type() != typeid(cryptonote::txin_to_key))
+      continue;
+    const txin_to_key& in_key = boost::get<txin_to_key>(tx.vin[i]);
+
+    // check if the key image belongs to us
+    const auto found = m_key_images.find(in_key.k_image);
+    if(found == m_key_images.end())
+    {
+      THROW_WALLET_EXCEPTION_IF(i > 0, error::wallet_internal_error, "subset of key images belong to us, very weird!");
+      LOG_PRINT_L0("This tx wasn't generated by this wallet");
+      return {};
+    }
+
+    // derive the real output keypair
+    const transfer_details& in_td = m_transfers[found->second];
+    const crypto::public_key in_tx_pubkey = get_tx_pub_key_from_extra(in_td.m_tx, in_td.m_pk_index);
+    keypair in_ephemeral;
+    crypto::key_image in_img;
+    THROW_WALLET_EXCEPTION_IF(!generate_key_image_helper(m_account.get_keys(), in_tx_pubkey, in_td.m_internal_output_index, in_ephemeral, in_img),
+      error::wallet_internal_error, "failed to generate key image");
+    THROW_WALLET_EXCEPTION_IF(in_key.k_image != in_img, error::wallet_internal_error, "key image mismatch");
+
+    // get output pubkeys in the ring
+    const std::vector<uint64_t> absolute_offsets = cryptonote::relative_output_offsets_to_absolute(in_key.key_offsets);
+    const size_t ring_size = in_key.key_offsets.size();
+    COMMAND_RPC_GET_OUTPUTS_BIN::request req = AUTO_VAL_INIT(req);
+    req.outputs.resize(ring_size);
+    for (size_t j = 0; j < ring_size; ++j)
+    {
+      req.outputs[j].amount = in_key.amount;
+      req.outputs[j].index = absolute_offsets[j];
+    }
+    COMMAND_RPC_GET_OUTPUTS_BIN::response res = AUTO_VAL_INIT(res);
+    m_daemon_rpc_mutex.lock();
```

Comment:
This style is done repeatedly in `wallet2` unfortunately - a { const boost::lock_guard<boost::mutex> lock{m_daemon_rpc_mutex}; ....; }` is helpful in case the invoke function throws.

Reply:
Thanks